Extractor/Connect/SessionManager.cs

```diff
@@ -0,0 +1,556 @@
+                        lastConnectionSwitch = DateTime.UtcNow;
+                    }
+                    // Attempt to update the service level, but only if we're not also reconnecting.
+                    // If we reconnect we fetch the service level anyway.
+                    else if (updateServiceLevel != null)
+                    {
+                        var res = await UpdateServiceLevel(updateServiceLevel.Value, false);
+                        if (res)
+                        {
+                            ConnectResult r;
+                            try
+                            {
+                                r = await connectionSource.Connect(connection, true, appConfig, token);
+                                lastConnectionSwitch = DateTime.UtcNow;
+                            }
+                            catch (Exception ex)
+                            {
+                                ExtractorUtils.LogException(log, ex, "Unexpected error attempting reconnect after service level change");
+                                state = SessionManagerState.Connecting;
+                                continue;
+                            }
+
+                            try
+                            {
+                                newConnection = await HandleConnectResult(r, token);
+                            }
+                            catch (Exception ex)
+                            {
+                                ExtractorUtils.LogException(log, ex, "Unexpected error when handling connect result after service level change");
+                                await CloseSession(r.Connection.Session, token);
+                                state = SessionManagerState.Connecting;
+                                continue;
+                            }
+                        }
+                    }
+                    if (newConnection && !isInitial)
+                    {
+                        // Don't run the callbacks inside the session manager loop, since
+                        // if we lose connection while the callbacks are running,
+                        // we need the loop to recover.
+                        client.Callbacks.TaskScheduler.ScheduleTask(null, _ => client.Callbacks.OnServerReconnect(client));
+                    }
+
+                    await onNewPendingActions.WaitAsync(token);
+                }
+            }
+        }
+
+        public static async Task<byte> ReadServiceLevel(ISession session, CancellationToken token)
+        {
+            var res = await session.ReadAsync(null, 0, TimestampsToReturn.Neither, new ReadValueIdCollection
+            {
+                new ReadValueId
+                {
+                    NodeId = VariableIds.Server_ServiceLevel,
+                    AttributeId = Attributes.Value
+                }
+            }, token);
+            var dv = res.Results[0];
```

Comment:
Is this guaranteed to not null-reference/index out of range?

Reply:
By the OPC-UA standard it is, but not by the SDK I think... We should throw if that happens, but I can throw something explicit.